File: 0132-palindrome-partitioning-ii/0132-palindrome-partitioning-ii.py
class Solution:
    def minCut(self, s: str) -> int:
        def ispal(s):
            return s[::]==s[::-1]
        def solve(s,i,j,dp):
            if i>=j:
                return 0
            if dp[i][j]!=-1:
                return dp[i][j]
            if ispal(s[i:j+1]):
                return 0
            minstep=math.inf
            for k in range(i,j):
                if ispal(s[i:k+1]):
                    temp=1+solve(s,k+1,j,dp)
                    minstep=min(minstep,temp)
            dp[i][j]=minstep
            return dp[i][j]
        DP=[[-1 for j in range(len(s))]for i in range(len(s))]
        return solve(s,0,len(s)-1,DP)
        

File: 0132-palindrome-partitioning-ii/0132-palindrome-partitioning-ii.cpp
class Solution {
public:
    int minCut(string s) {
        int n=s.size();
        vector<vector<bool>>ispal(n,vector<bool>(n,false));
        for(int i=0;i<n;i++)
        {
            int left=i,right=i+1;
            while(left>=0 and right<n and s[left]==s[right])
            {
                ispal[left][right]=true;
                left--;right++;
            }
            left=i;right=i;
            while(left>=0 and right<n and s[left]==s[right])
            {
                ispal[left][right]=true;
                left--;right++;
            }
        }
        unordered_map<int,int>DP;
        return solve(0,s,ispal,DP)-1;
    }
private:
    int solve(int ind,string &s,vector<vector<bool>>&ispal,unordered_map<int,int>&DP)
    {
        if(ind==s.size())return 0;
        else
        {
            if(DP.find(ind)!=DP.end())return DP[ind];
            int ans=INT_MAX;
            for(int i=ind;i<s.size();i++)
            {
                if(ispal[ind][i])
                {
                    ans=min(ans,1+solve(i+1,s,ispal,DP));
                }
            }
            return DP[ind]=ans;
        }
    }
};